### PR TITLE
WIP: gpg signing (followup of #420)

### PR DESCRIPTION
**Pull request by magiconair**
* gpg_signing -> sign
* optionally sign checksum file only
* make signature file extension confiugrable
* support encrypted signing keys

This is based on PR #420 with some additional features. 

Tests and docs are still missing.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] I have read the **CONTRIBUTING** document.
- [ ] `make ci` passes on my machine.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 546fb49474c75c333236760715cacf6a39fa759d
  * Your commit message does not contain a body
  * Commits must be in the following format: **%{type}: %{description}**

 * Commit: 4be678dc7c9a2aa8dfa001054cf7c5e267a201f2
  * Your commit message does not contain a body

 * Commit: dec8ff37b91cf1363262d771954a339969deb30d
  * Your commit message does not contain a body

Guidelines are available at https://github.com/goreleaser/goreleaser/blob/master/CONTRIBUTING.md#create-a-commit

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 546fb49474c75c333236760715cacf6a39fa759d
  * Your commit message does not contain a body
  * Commits must be in the following format: **%{type}: %{description}**

 * Commit: 4be678dc7c9a2aa8dfa001054cf7c5e267a201f2
  * Your commit message does not contain a body

 * Commit: dec8ff37b91cf1363262d771954a339969deb30d
  * Your commit message does not contain a body

Guidelines are available at https://github.com/goreleaser/goreleaser/blob/master/CONTRIBUTING.md#create-a-commit

---
This message was auto-generated by https://gitcop.com

**Comment by caarlos0** [comment]
overall it looks good - the build failed though.

also: mind adding some unit tests?

**Comment by magiconair** [comment]
Yes. I’ll clean that up.

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 546fb49474c75c333236760715cacf6a39fa759d
  * Your commit message does not contain a body
  * Commits must be in the following format: **%{type}: %{description}**

 * Commit: 4be678dc7c9a2aa8dfa001054cf7c5e267a201f2
  * Your commit message does not contain a body

 * Commit: dec8ff37b91cf1363262d771954a339969deb30d
  * Your commit message does not contain a body

Guidelines are available at https://github.com/goreleaser/goreleaser/blob/master/CONTRIBUTING.md#create-a-commit

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 546fb49474c75c333236760715cacf6a39fa759d
  * Your commit message does not contain a body
  * Commits must be in the following format: **%{type}: %{description}**

 * Commit: 4be678dc7c9a2aa8dfa001054cf7c5e267a201f2
  * Your commit message does not contain a body

 * Commit: dec8ff37b91cf1363262d771954a339969deb30d
  * Your commit message does not contain a body

Guidelines are available at https://github.com/goreleaser/goreleaser/blob/master/CONTRIBUTING.md#create-a-commit

---
This message was auto-generated by https://gitcop.com

File: pipeline/sign/sign.go [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+	if len(rawKeyID) != 16 {
+		return fmt.Errorf("invalid gpg_key_id %q, needs to be a 8 byte long hex key id", rawKeyID)
+	}
+
+	keyID, err := strconv.ParseUint(rawKeyID, 16, 64)
+	if err != nil {
+		return fmt.Errorf("invalid gpg_key_id '%s': %s", rawKeyID, err)
+	}
+
+	// todo(fs): is this always in that location?
```
Comment (caarlos0) [inline pipeline/sign/sign.go]: no... it can be Defined using the `secret-keyring <dir>/secring.gpg` config...

https://www.gnupg.org/gph/en/manual/r1593.html

Comment (caarlos0) [reply pipeline/sign/sign.go]: but I think we can deal with this when someone needs it :)

File: context/context.go [hunk 2]
```diff
@@ -37,6 +37,7 @@ type Context struct {
 	Git          GitInfo
 	Binaries     map[string]map[string][]Binary
 	Artifacts    []string
+	Checksums    []string
```
Comment (caarlos0) [inline context/context.go]: why is this needed?

Comment (magiconair) [reply context/context.go]: So that you can only sign the checksum file

**Comment by magiconair** [comment]
I’ll add tests and docs tonight.

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: bec5be352aa6286edc2f9502591cdff6eacf34a7
  * Your commit message body contains a line that is longer than 72 characters

 * Commit: 2fee52707426a3507a9db461bfe019643016b873
  * Your commit message does not contain a body
  * Commits must be in the following format: **%{type}: %{description}**

Guidelines are available at https://github.com/goreleaser/goreleaser/blob/master/CONTRIBUTING.md#create-a-commit

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: bec5be352aa6286edc2f9502591cdff6eacf34a7
  * Your commit message body contains a line that is longer than 72 characters

 * Commit: 2fee52707426a3507a9db461bfe019643016b873
  * Your commit message does not contain a body
  * Commits must be in the following format: **%{type}: %{description}**

Guidelines are available at https://github.com/goreleaser/goreleaser/blob/master/CONTRIBUTING.md#create-a-commit

---
This message was auto-generated by https://gitcop.com

**Comment by magiconair** [comment]
I'm working on test cases but hit a road-block which makes me question the approach in general. The `secring.gpg` file which `golang.org/x/crypto/opengpg` can read is no longer used by GnuPG >= 2.1. 

I might be able to find a way to make this work but I think a better approach would be to configure a `sign_cmd` which signs a single file and then that can be anything you like. This makes management of keyring paths and gpg options unnecessary. We'd inject the name of the file to sign as `${f}` so that you can template the command. `os.Expand` makes that simple.

```yaml
sign:
  sign_cmd: "gpg --output fabio-${f}.sig --detach-sig ${f}
  artifacts: checksum # (all|checksum)
```

I'll try that in a separate branch.

**Comment by caarlos0** [comment]
@magiconair sounds like a good idea to me!

**Comment by magiconair** [comment]
Abandoing in favor of #460
